src/map/skill.c

```diff
@@ -7708,24 +7708,31 @@ static int skill_castend_nodamage_id(struct block_list *src, struct block_list *
 		case MO_ABSORBSPIRITS:
 		{
 			int sp = 0;
+			bool success = false;
 			if (dstsd != NULL && dstsd->spiritball != 0
 			 && (sd == dstsd || map_flag_vs(src->m) || (sd && sd->duel_group && sd->duel_group == dstsd->duel_group))
 			 && (dstsd->job & MAPID_BASEMASK) != MAPID_GUNSLINGER
 			 ) {
```

Comment:
I think some of those checks could be cleaned up because they are being checked in `skill_check_condition_castend` now.

But in the other hand, `skill_check_condition_castend` has a few checks for autocast at the beginning that could allow the "invalid skill" to get here... so I am not sure if it is safe to clear the checks here.

Reply:
those: (reformatted for comments)
```C
if (
    /* Only for player targets */
    dstsd != NULL
    /* AND must have spheres (1) */
    && dstsd->spiritball != 0
    /* AND must be (self OR enemy) (2) */
    && (sd == dstsd || map_flag_vs(src->m) || (sd && sd->duel_group && sd->duel_group == dstsd->duel_group))
    /* AND NOT gunslinger (3) */
    && (dstsd->job & MAPID_BASEMASK) != MAPID_GUNSLINGER
) {
       /* Skill works */
}
```

from the updated `skill_check_condition_castend` (when not being from auto cast):
```
case MO_ABSORBSPIRITS: {
	struct map_session_data *tgtsd = BL_CAST(BL_PC, target);
	// Non-player targets would just continue
	if (tgtsd == NULL)
 		break;

	// Fail if 0 spheres (1)
	// Fail if gunslinger (3)
	if (tgtsd->spiritball == 0 
	    || (tgtsd->job & MAPID_BASEMASK) == MAPID_GUNSLINGER) 
		/* Fail */

	// Fail if using in a friend (not self) (2)
	if (sd != tgtsd
	    && battle->check_target(&sd->bl, &tgtsd->bl, BCT_NOENEMY) == 1)
		/* Fail*/

	/* Pass */
```

so in a normal skill usage, used against a `sd`, we would only get here when:
1. target is not gunslinger
2. target has 1 or more spheres
3. target is an enemy or self

so both checks are equivalent and simply checking if `dstsd != NULL` should be enough.

but we have these lines for auto cast: https://github.com/HerculesWS/Hercules/pull/3215/files#diff-759fea9dcd40522e86217dd352aadd687488dff49c67aa880610d4aeaaeb89c0R16515-R16521

Reply:
I see what you mean. I agree that some of the check are redundant, but I wouldn't remove all of them since I believe that some here have a slightly different purpose, especially the gunslinger one.

In `skill_check_condition_castend()` they're deciding whether to make the skill fail entirely, while here they're preventing unintended side effects caused by the shared implementation between spirit spheres and gunslinger coins.

Let's say that tomorrow we introduce a new SC (or, alternatively, a plugin introduces a new SC) that makes the caster bypass requirements and cast skills unconditionately, bypassing code in `skill_check_condition_castend`. If the gunslinger check is removed from here, the skill would have a different, implementation-defined, behavior consuming coins even if it's supposed to consume spirit spheres. I'd feel like that behavior is a bug caused by the missing check here, and the correct behavior would be for the skill to have no effect since there are no spheres to remove.

The vs check though can be removed from here, since it isn't protecting implementation details that would break the skill's behavior, but simply implementing a game mechanic that imposed the skill should only be used on friendly units.

I believe that each piece of code should not rely on the way it is currently called, and should assume that plugins or future changes may call it differently, as long as they don't do anything illegal. And it should offer protection against illegal inputs that could lead to unexpected behavior or crashes (and document what's illegal) through assertions.

Reply:
that's a very good point of view! thank you, Haru.
I have update the commit remove the vs check only